nameko/dependencies.py

```diff
@@ -68,75 +69,71 @@ def kill(self):
-        the worker spawn.
-
-        DependencyProviders should do any post-processing here, raising
-        exceptions in the event of failure.
-
-        Example: a database session provider may commit the session
-
-        Args:
-            - worker_ctx: see
-                ``nameko.containers.ServiceContainer.spawn_worker``
-        """
-
     def bind(self, name, container):
-        """ Bind this DependencyProvider instance to ``container`` using the
-        given ``name`` to identify the resource on the hosted service.
-
-        Called during ServiceContainer initialisation. The DependencyProvider
-        instance is created and then bound to the ServiceContainer instance
-        controlling its lifecyle.
+        """ Bind a clone of this extension to `container` with the given
+        `name`.
         """
-        self.name = name
-        self.container = container
-        self.bound = True
+        if self.__clone:  # TODO: prefer this over check in clone?
+            raise RuntimeError('Cloned extensions cannot be bound.')
+
+        # if this extension is shared and there's already a cloned instance,
+        # return that
+        if self.__shared:
+            shared_extensions.setdefault(container, {})
+            shared = shared_extensions[container].get(self.sharing_key)
+            if shared:
+                return shared
+
+        # clone this extension and associate it with the container
+        clone = self.clone()
+        clone.name = name
+        clone.container = container
+
+        # recursively bind nested extensions
+        for ext_name, ext in inspect.getmembers(self, is_extension):
+            setattr(clone, ext_name, ext.bind(ext_name, container))
+
+        # if this extension is shared, save the cloned instance
+        if self.__shared:
+            shared_extensions[container][self.sharing_key] = clone
+
+        return clone
+
+    def clone(self):
+        if self.__state is None:
+            raise RuntimeError('Undefined extension state. Did you forget to '
+                               'call super().__init__() in your subclass?')
+
+        if self.__clone:  # TODO: need this even with the check in bind above?
+            raise RuntimeError('Cloned extensions cannot be cloned.')
+
+        cls = type(self)
+        args, kwargs = self.__state
+        instance = cls(*args, **kwargs)
+        instance.__clone = True
+        # alternative: del clone.clone; del clone.bind
+        return instance
 
     @property
-    def nested_dependencies(self):
-        """ Recusively yield nested dependencies of DependencyProvider.
-
-        For example, with an instance of `:class:~nameko.rpc.RpcProvider`
-        called ``rpc``::
-
-            >>> deps = list(rpc.nested_dependencies)
-            >>> deps
-            [<nameko.rpc.RpcConsumer object at 0x10d125690>,
-             <nameko.messaging.QueueConsumer object at 0x10d5b8f50>]
-            >>>
+    def extensions(self):
```

Comment:
could we have a more descriptive name here? (not sure what yet) 

(esp. given that we include self)